<a href="https://colab.research.google.com/github/akshayabhale1406/SCLR_Delhivery_FE/blob/main/Delhivery_FE.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Business Case: Delhivery - Feature Engineering**

## Problem Statement : About Delhivery

Delhivery is the largest and fastest-growing fully integrated player in India by revenue in Fiscal 2021. They aim to build the operating system for commerce, through a combination of world-class infrastructure, logistics operations of the highest quality, and cutting-edge engineering and technology capabilities.

The Data team builds intelligence and capabilities using this data that helps them to widen the gap between the quality, efficiency, and profitability of their business versus their competitors.

### Importing the relevant libraries

In [986]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns
pd.set_option('display.max_columns', None)

### Loading CSV as Dataset

In [987]:
df = pd.read_csv("/content/sample_data/delhivery_data.csv")

### Checking

In [988]:
df.head()

data          trip_creation_time  \
0  training  2018-09-20 02:35:36.476840   
1  training  2018-09-20 02:35:36.476840   
2  training  2018-09-20 02:35:36.476840   
3  training  2018-09-20 02:35:36.476840   
4  training  2018-09-20 02:35:36.476840   

                                 route_schedule_uuid route_type  \
0  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
1  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
2  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
3  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
4  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   

                 trip_uuid source_center                 source_name  \
0  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
1  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
2  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
3  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
4  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   

  destination_center               destination_name  \
0       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
1       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
2       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
3       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
4       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   

                od_start_time                 od_end_time  \
0  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   
1  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   
2  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   
3  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   
4  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   

   start_scan_to_end_scan  is_cutoff  cutoff_factor  \
0                    86.0       True              9   
1                    86.0       True             18   
2                    86.0       True             27   
3                    86.0       True             36   
4                    86.0      False             39   

             cutoff_timestamp  actual_distance_to_destination  actual_time  \
0         2018-09-20 04:27:55                       10.435660         14.0   
1         2018-09-20 04:17:55                       18.936842         24.0   
2  2018-09-20 04:01:19.505586                       27.637279         40.0   
3         2018-09-20 03:39:57                       36.118028         62.0   
4         2018-09-20 03:33:55                       39.386040         68.0   

   osrm_time  osrm_distance    factor  segment_actual_time  segment_osrm_time  \
0       11.0        11.9653  1.272727                 14.0               11.0   
1       20.0        21.7243  1.200000                 10.0                9.0   
2       28.0        32.5395  1.428571                 16.0                7.0   
3       40.0        45.5620  1.550000                 21.0               12.0   
4       44.0        54.2181  1.545455                  6.0                5.0   

   segment_osrm_distance  segment_factor  
0                11.9653        1.272727  
1                 9.7590        1.111111  
2                10.8152        2.285714  
3                13.0224        1.750000  
4                 3.9153        1.200000

### Checking the shape of the data

In [989]:
df.shape

(144867, 24)

### Checking column name

In [990]:
df.columns

Index(['data', 'trip_creation_time', 'route_schedule_uuid', 'route_type',
       'trip_uuid', 'source_center', 'source_name', 'destination_center',
       'destination_name', 'od_start_time', 'od_end_time',
       'start_scan_to_end_scan', 'is_cutoff', 'cutoff_factor',
       'cutoff_timestamp', 'actual_distance_to_destination', 'actual_time',
       'osrm_time', 'osrm_distance', 'factor', 'segment_actual_time',
       'segment_osrm_time', 'segment_osrm_distance', 'segment_factor'],
      dtype='object')

**OBSERVATION**: This data consists of 144867 checkpoints described across 24 columns. The columns are as follows:
1) data
2) trip_creation_time
3) route_schedule_uuid
4) route_type
5) trip_uuid
6) source_center
7) source_name
8) destination_center
9) destination_name
10) od_start_time
11) od_end_time
12) start_scan_to_end_scan
13) is_cutoff
14) cutoff_factor
15) cutoff_timestamp
16) actual_distance_to_destination
17) actual_time
18) osrm_time
19) osrm_distance
20) factor
21) segment_actual_time
22) segment_osrm_time
23) segment_osrm_distance


### Checking data type of each column

In [991]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144867 entries, 0 to 144866
Data columns (total 24 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   data                            144867 non-null  object 
 1   trip_creation_time              144867 non-null  object 
 2   route_schedule_uuid             144867 non-null  object 
 3   route_type                      144867 non-null  object 
 4   trip_uuid                       144867 non-null  object 
 5   source_center                   144867 non-null  object 
 6   source_name                     144574 non-null  object 
 7   destination_center              144867 non-null  object 
 8   destination_name                144606 non-null  object 
 9   od_start_time                   144867 non-null  object 
 10  od_end_time                     144867 non-null  object 
 11  start_scan_to_end_scan          144867 non-null  float64
 12  is_cutoff       

**OBSERVATION**: The time column should be of correct datatype, that is: datetime

### Convert the od_start_time and od_end_time to datetime data type

In [992]:
df["trip_creation_time"] = pd.to_datetime(df["trip_creation_time"])

df["od_start_time"] = pd.to_datetime(df["od_start_time"])
df["od_end_time"] = pd.to_datetime(df["od_end_time"])

In [993]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144867 entries, 0 to 144866
Data columns (total 24 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   data                            144867 non-null  object        
 1   trip_creation_time              144867 non-null  datetime64[ns]
 2   route_schedule_uuid             144867 non-null  object        
 3   route_type                      144867 non-null  object        
 4   trip_uuid                       144867 non-null  object        
 5   source_center                   144867 non-null  object        
 6   source_name                     144574 non-null  object        
 7   destination_center              144867 non-null  object        
 8   destination_name                144606 non-null  object        
 9   od_start_time                   144867 non-null  datetime64[ns]
 10  od_end_time                     144867 non-null  datetim

### Checking for missing vaules in column

In [994]:
df.isnull().sum()

data                                0
trip_creation_time                  0
route_schedule_uuid                 0
route_type                          0
trip_uuid                           0
source_center                       0
source_name                       293
destination_center                  0
destination_name                  261
od_start_time                       0
od_end_time                         0
start_scan_to_end_scan              0
is_cutoff                           0
cutoff_factor                       0
cutoff_timestamp                    0
actual_distance_to_destination      0
actual_time                         0
osrm_time                           0
osrm_distance                       0
factor                              0
segment_actual_time                 0
segment_osrm_time                   0
segment_osrm_distance               0
segment_factor                      0
dtype: int64

**OBSERVATION**: There are missing values in  columns: source_name, destination_name, and segment factor. Since the missing values are less than 1% of the total data, we can directly drop the NULL values.



### Dropping the NULL values from the dataset

In [995]:
df = df.dropna(how ='any')

### Checking the shape of data after dropping NULL Values

In [996]:
df.shape

(144316, 24)

**OBSERVATION**: The new dataset is decreased by less than 1% of the actual raw dataset. This is acceptable for further Data Analysis.

In [997]:
df_null_check = df.isnull().sum()

### Checking to see if there are any NULL values in the new dataset.

In [998]:
df_null_check

data                              0
trip_creation_time                0
route_schedule_uuid               0
route_type                        0
trip_uuid                         0
source_center                     0
source_name                       0
destination_center                0
destination_name                  0
od_start_time                     0
od_end_time                       0
start_scan_to_end_scan            0
is_cutoff                         0
cutoff_factor                     0
cutoff_timestamp                  0
actual_distance_to_destination    0
actual_time                       0
osrm_time                         0
osrm_distance                     0
factor                            0
segment_actual_time               0
segment_osrm_time                 0
segment_osrm_distance             0
segment_factor                    0
dtype: int64

**OBSERVATION**: No NULL values in the new dataset.

### Checking if any orders are returned to source location.

In [999]:
returned_orders = df.groupby("trip_uuid").apply(lambda group: pd.DataFrame({"Returned_Packages": [any(group["source_center"] == group["destination_center"])]})).reset_index()
returned_orders.head(20)

<ipython-input-999-2186fcab8b80>:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  returned_orders = df.groupby("trip_uuid").apply(lambda group: pd.DataFrame({"Returned_Packages": [any(group["source_center"] == group["destination_center"])]})).reset_index()


trip_uuid  level_1  Returned_Packages
0   trip-153671041653548748        0              False
1   trip-153671042288605164        0              False
2   trip-153671043369099517        0              False
3   trip-153671046011330457        0              False
4   trip-153671052974046625        0              False
5   trip-153671055416136166        0              False
6   trip-153671066201138152        0              False
7   trip-153671066826362165        0              False
8   trip-153671074033284934        0              False
9   trip-153671079956500691        0              False
10  trip-153671090980523004        0              False
11  trip-153671110078355292        0              False
12  trip-153671117375890592        0              False
13  trip-153671121411074590        0              False
14  trip-153671122262300031        0              False
15  trip-153671127071609610        0              False
16  trip-153671128962918389        0              False
17  trip-153671143043841452        0              False
18  trip-153671173668736946        0              False
19  trip-153671174968648046        0              False

In [1000]:
true_instances = returned_orders[returned_orders['Returned_Packages'] == True].index.tolist()
true_instances


[]

**OBSERVATION**: The below shows if any values are TRUE for Returned_Packages. Since the returned list is empty, it means no packages were returned.

### The current details are splitted for different trips as per different check-points for that trips. Data can be analysed better if the times of each check-points can be summed up for each trip ID.

In [1001]:
df_combined_times = df.groupby("trip_uuid")[["segment_osrm_time", "segment_actual_time", "segment_osrm_distance"]].sum()
df_combined_times["segment_osrm_distance"] = np.round(df_combined_times["segment_osrm_distance"], 2)
df_sorted_times = df_combined_times.sort_values(by = "segment_actual_time", ascending = True).reset_index()
df_sorted_times.head(50)

trip_uuid  segment_osrm_time  segment_actual_time  \
0   trip-153807427473576350                8.0                  9.0   
1   trip-153806863340281809                7.0                  9.0   
2   trip-153803888649276314                7.0                 10.0   
3   trip-153764261916618451                8.0                 10.0   
4   trip-153855687660912584                7.0                 10.0   
5   trip-153780834256270168               12.0                 10.0   
6   trip-153725458328238387               12.0                 11.0   
7   trip-153803758183582299               12.0                 11.0   
8   trip-153757853452165881               12.0                 11.0   
9   trip-153716585040807264               11.0                 11.0   
10  trip-153766197332255140                8.0                 11.0   
11  trip-153707928006502698               10.0                 11.0   
12  trip-153854182124368611               12.0                 12.0   
13  trip-153728911454561981               11.0                 12.0   
14  trip-153786579893071545               11.0                 12.0   
15  trip-153816453586757131               12.0                 12.0   
16  trip-153842335872450338               12.0                 12.0   
17  trip-153765458951431978               16.0                 12.0   
18  trip-153705162529056762               12.0                 12.0   
19  trip-153722896880816997                7.0                 12.0   
20  trip-153679789619211844                7.0                 12.0   
21  trip-153713819335800961               11.0                 13.0   
22  trip-153815511086978641                9.0                 13.0   
23  trip-153850119815065023               10.0                 13.0   
24  trip-153838461655152916                9.0                 13.0   
25  trip-153768237859208805                7.0                 13.0   
26  trip-153738218098136608               13.0                 13.0   
27  trip-153821245210601604                9.0                 13.0   
28  trip-153738889579049016               11.0                 13.0   
29  trip-153699387226763684                7.0                 13.0   
30  trip-153847034707448764                9.0                 13.0   
31  trip-153721738031711906               11.0                 13.0   
32  trip-153829835420221745                7.0                 13.0   
33  trip-153791504799756600               12.0                 13.0   
34  trip-153752218422279635               11.0                 13.0   
35  trip-153743449450488839               11.0                 13.0   
36  trip-153765583943264913                9.0                 13.0   
37  trip-153808848921752145               15.0                 13.0   
38  trip-153746554953775475               11.0                 13.0   
39  trip-153852083210089878               14.0                 13.0   
40  trip-153750504970744403               12.0                 13.0   
41  trip-153692300567549463               11.0                 14.0   
42  trip-153830066027083341                8.0                 14.0   
43  trip-153832808141107171                9.0                 14.0   
44  trip-153803995207530367               11.0                 14.0   
45  trip-153821428331375306               10.0                 14.0   
46  trip-153673212880241423                7.0                 14.0   
47  trip-153739602851447154               16.0                 14.0   
48  trip-153787176752521342               12.0                 14.0   
49  trip-153748365439419345               12.0                 14.0   

    segment_osrm_distance  
0                   10.14  
1                   10.11  
2                    9.39  
3                    9.33  
4                    9.40  
5                   13.10  
6                    9.66  
7                    9.78  
8                   10.18  
9                   15.26  
10                   9.87  
11                  14.48  
12                   9.96  
13  

### Calculation of time difference between OSRM time and Actual Time to know how much time error present present in the OSRM time calculation.

In [1002]:
df_sorted_times["time_calc_error"] = df_sorted_times["segment_osrm_time"] - df_sorted_times["segment_actual_time"]
df_sorted_times.head()

trip_uuid  segment_osrm_time  segment_actual_time  \
0  trip-153807427473576350                8.0                  9.0   
1  trip-153806863340281809                7.0                  9.0   
2  trip-153803888649276314                7.0                 10.0   
3  trip-153764261916618451                8.0                 10.0   
4  trip-153855687660912584                7.0                 10.0   

   segment_osrm_distance  time_calc_error  
0                  10.14             -1.0  
1                  10.11             -2.0  
2                   9.39             -3.0  
3                   9.33             -2.0  
4                   9.40             -3.0

In [1003]:
df_sorted_tce = df_sorted_times.sort_values(by = "time_calc_error", ascending = True)
df_sorted_tce.head()

trip_uuid  segment_osrm_time  segment_actual_time  \
14786  trip-153840656812932039             2201.0               6230.0   
14785  trip-153805929645698249             2494.0               5768.0   
14777  trip-153736469866480991             1123.0               4129.0   
14784  trip-153814369747664517             2510.0               5427.0   
14780  trip-153724675612503042             1667.0               4504.0   

       segment_osrm_distance  time_calc_error  
14786                3150.65          -4029.0  
14785                3418.18          -3274.0  
14777                1611.54          -3006.0  
14784                3456.04          -2917.0  
14780                2323.33          -2837.0

In [1004]:
df_sorted_tce = df_sorted_times.sort_values(by = "time_calc_error", ascending = False)
df_sorted_tce.head()

trip_uuid  segment_osrm_time  segment_actual_time  \
8837  trip-153677882741378033              418.0                208.0   
8618  trip-153686822090217892              363.0                197.0   
2288  trip-153751592030055181              194.0                 50.0   
8889  trip-153721065936066058              317.0                210.0   
8430  trip-153711137119449163              285.0                188.0   

      segment_osrm_distance  time_calc_error  
8837                 552.59            210.0  
8618                 407.84            166.0  
2288                 134.43            144.0  
8889                 408.10            107.0  
8430                 374.64             97.0

In [1005]:
first_od_start_times = df.groupby('trip_uuid')['od_start_time'].first()
df_combined_times = df_combined_times.merge(first_od_start_times, left_index=True, right_index=True).reset_index()
df_combined_times["trip_start_day"] = df_combined_times["od_start_time"].dt.day
df_combined_times["trip_start_month"] = df_combined_times["od_start_time"].dt.month
df_combined_times["trip_start_year"] = df_combined_times["od_start_time"].dt.year
trip_start_location = df.groupby('trip_uuid')['source_name'].first()
trip_end_location = df.groupby('trip_uuid')['destination_name'].last()
df_combined_times = df_combined_times.set_index('trip_uuid').merge(trip_start_location, left_index=True, right_index=True).reset_index()
df_combined_times = df_combined_times.set_index('trip_uuid').merge(trip_end_location, left_index=True, right_index=True).reset_index()
df_combined_times = df_combined_times.drop(columns=['od_start_time'])
df_combined_times.head()

trip_uuid  segment_osrm_time  segment_actual_time  \
0  trip-153671041653548748             1008.0               1548.0   
1  trip-153671042288605164               65.0                141.0   
2  trip-153671043369099517             1941.0               3308.0   
3  trip-153671046011330457               16.0                 59.0   
4  trip-153671052974046625              115.0                340.0   

   segment_osrm_distance  trip_start_day  trip_start_month  trip_start_year  \
0                1320.47              12                 9             2018   
1                  84.19              12                 9             2018   
2                2545.27              12                 9             2018   
3                  19.88              12                 9             2018   
4                 146.79              12                 9             2018   

                          source_name                   destination_name  
0  Bhopal_Trnsport_H (Madhya Pradesh)      Gurgaon_Bilaspur_HB (Haryana)  
1       Tumkur_Veersagr_I (Karnataka)  Chikblapur_ShntiSgr_D (Karnataka)  
2    Bangalore_Nelmngla_H (Karnataka)     Chandigarh_Mehmdpur_H (Punjab)  
3            Mumbai Hub (Maharashtra)     Mumbai_MiraRd_IP (Maharashtra)  
4              Bellary_Dc (Karnataka)             Bellary_Dc (Karnataka)

### Creating a column for Starting Point and Ending Point of a trip using States names

In [1006]:
df_combined_times["source_name"] = df_combined_times["source_name"].str.split("(").str[1]
df_combined_times["source_name"] = df_combined_times["source_name"].str.split(")").str[0]
df_combined_times["source_name"] = df_combined_times["source_name"].apply(lambda x : x.upper())
df_combined_times["destination_name"] = df_combined_times["destination_name"].str.split("(").str[1]
df_combined_times["destination_name"] = df_combined_times["destination_name"].str.split(")").str[0]
df_combined_times["destination_name"] = df_combined_times["destination_name"].apply(lambda x : x.upper())
df_combined_times

trip_uuid  segment_osrm_time  segment_actual_time  \
0      trip-153671041653548748             1008.0               1548.0   
1      trip-153671042288605164               65.0                141.0   
2      trip-153671043369099517             1941.0               3308.0   
3      trip-153671046011330457               16.0                 59.0   
4      trip-153671052974046625              115.0                340.0   
...                        ...                ...                  ...   
14782  trip-153861095625827784               62.0                 82.0   
14783  trip-153861104386292051               11.0                 21.0   
14784  trip-153861106442901555               88.0                281.0   
14785  trip-153861115439069069              221.0                258.0   
14786  trip-153861118270144424               67.0                274.0   

       segment_osrm_distance  trip_start_day  trip_start_month  \
0                    1320.47              12                 9   
1                      84.19              12                 9   
2                    2545.27              12                 9   
3                      19.88              12                 9   
4                     146.79              12                 9   
...                      ...             ...               ...   
14782                  64.86               3                10   
14783                  16.09               3                10   
14784                 104.89               3                10   
14785                 223.53               3                10   
14786                  80.58               4                10   

       trip_start_year     source_name destination_name  
0                 2018  MADHYA PRADESH          HARYANA  
1                 2018       KARNATAKA        KARNATAKA  
2                 2018       KARNATAKA           PUNJAB  
3                 2018     MAHARASHTRA      MAHARASHTRA  
4                 2018       KARNATAKA        KARNATAKA  
...                ...             ...              ...  
14782             2018          PUNJAB           PUNJAB  
14783             2018         HARYANA          HARYANA  
14784             2018   UTTAR PRADESH    UTTAR PRADESH  
14785             2018      TAMIL NADU       TAMIL NADU  
14786             2018       KARNATAKA        KARNATAKA  

[14787 rows x 9 columns]

In [1007]:
od_start = df.groupby('trip_uuid')['od_start_time'].first()
od_end = df.groupby('trip_uuid')['od_end_time'].last()
df_combined_times = df_combined_times.set_index('trip_uuid').merge(od_start, left_index=True, right_index=True).reset_index()
df_combined_times = df_combined_times.set_index('trip_uuid').merge(od_end, left_index=True, right_index=True).reset_index()
df_combined_times["trip_duration"] = (df_combined_times["od_end_time"] - df_combined_times["od_start_time"]).dt.total_seconds()/60
df_combined_times.head()


trip_uuid  segment_osrm_time  segment_actual_time  \
0  trip-153671041653548748             1008.0               1548.0   
1  trip-153671042288605164               65.0                141.0   
2  trip-153671043369099517             1941.0               3308.0   
3  trip-153671046011330457               16.0                 59.0   
4  trip-153671052974046625              115.0                340.0   

   segment_osrm_distance  trip_start_day  trip_start_month  trip_start_year  \
0                1320.47              12                 9             2018   
1                  84.19              12                 9             2018   
2                2545.27              12                 9             2018   
3                  19.88              12                 9             2018   
4                 146.79              12                 9             2018   

      source_name destination_name              od_start_time  \
0  MADHYA PRADESH          HARYANA 2018-09-12 00:00:16.535741   
1       KARNATAKA        KARNATAKA 2018-09-12 00:00:22.886430   
2       KARNATAKA           PUNJAB 2018-09-12 00:00:33.691250   
3     MAHARASHTRA      MAHARASHTRA 2018-09-12 00:01:00.113710   
4       KARNATAKA        KARNATAKA 2018-09-12 00:02:09.740725   

                 od_end_time  trip_duration  
0 2018-09-13 13:40:23.123744    2260.109800  
1 2018-09-12 03:01:59.598855     181.611874  
2 2018-09-14 17:34:55.442454    3934.362520  
3 2018-09-12 01:41:29.809822     100.494935  
4 2018-09-12 12:00:30.683231     718.349042

In [1008]:
# combined_scan_time = df.groupby("")

In [1009]:
df_combined_times = df_combined_times.drop(columns=['od_start_time', 'od_end_time'])
start_scan_to_end_scan = df.groupby(["trip_uuid","destination_center"])["start_scan_to_end_scan"].first()
start_scan_to_end_scan = start_scan_to_end_scan.reset_index()
start_scan_to_end_scan.head()

trip_uuid destination_center  start_scan_to_end_scan
0  trip-153671041653548748       IND000000ACB                  1260.0
1  trip-153671041653548748       IND209304AAA                   999.0
2  trip-153671042288605164       IND561203AAB                   122.0
3  trip-153671042288605164       IND562101AAA                    58.0
4  trip-153671043369099517       IND000000ACB                  3099.0

In [1010]:
df_combined_times1 = start_scan_to_end_scan.groupby("trip_uuid")["start_scan_to_end_scan"].sum()
df_combined_times1 = df_combined_times1.reset_index()
df_combined_times1.head()

trip_uuid  start_scan_to_end_scan
0  trip-153671041653548748                  2259.0
1  trip-153671042288605164                   180.0
2  trip-153671043369099517                  3933.0
3  trip-153671046011330457                   100.0
4  trip-153671052974046625                   717.0

In [1011]:
# df_combined_times = df_combined_times.set_index('trip_uuid').merge(start_scan_to_end_scan, left_index=True, right_index=True).reset_index()
# df_combined_times.head()

In [1012]:
df_combined_times = df_combined_times.merge(df_combined_times1, on = "trip_uuid", how = "left")
df_combined_times.head()

trip_uuid  segment_osrm_time  segment_actual_time  \
0  trip-153671041653548748             1008.0               1548.0   
1  trip-153671042288605164               65.0                141.0   
2  trip-153671043369099517             1941.0               3308.0   
3  trip-153671046011330457               16.0                 59.0   
4  trip-153671052974046625              115.0                340.0   

   segment_osrm_distance  trip_start_day  trip_start_month  trip_start_year  \
0                1320.47              12                 9             2018   
1                  84.19              12                 9             2018   
2                2545.27              12                 9             2018   
3                  19.88              12                 9             2018   
4                 146.79              12                 9             2018   

      source_name destination_name  trip_duration  start_scan_to_end_scan  
0  MADHYA PRADESH          HARYANA    2260.109800                  2259.0  
1       KARNATAKA        KARNATAKA     181.611874                   180.0  
2       KARNATAKA           PUNJAB    3934.362520                  3933.0  
3     MAHARASHTRA      MAHARASHTRA     100.494935                   100.0  
4       KARNATAKA        KARNATAKA     718.349042                   717.0